testv2/e2e/cloudprovider.go

```diff
@@ -0,0 +1,342 @@
+/*
+Copyright 2022 The KubeOne Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package e2e
+
+import (
+	"context"
+	"fmt"
+	"testing"
+	"time"
+
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/wait"
+	ctrlruntimeclient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func supportsStorage(infra Infra) bool {
+	switch infra.Provider() {
+	case "aws", "digitalocen", "hetzner", "gce", "openstack":
+		return true
+	default:
+		return false
+	}
+}
+
+func supportsLoadBalancer(infra Infra) bool {
+	switch infra.Provider() {
+	case "aws", "digitalocean", "hetzner", "gce", "openstack":
+		return true
+	default:
+		return false
+	}
+}
+
+const (
+	namespaceName   = "test-cloud-provider"
+	statefulsetName = "test-sts"
+	serviceName     = "test-svc"
+
+	testPollPeriod = 10 * time.Second
+	testTimeout    = 10 * time.Minute
+)
+
+var (
+	cloudProviderLabels = map[string]string{"app": "test-cp"}
+)
+
+type cloudProviderTests struct {
+	ctx    context.Context
+	client ctrlruntimeclient.Client
+	infra  Infra
+}
+
+func newCloudProviderTests(client ctrlruntimeclient.Client, infra Infra) *cloudProviderTests {
+	return &cloudProviderTests{
+		ctx:    context.Background(),
+		client: client,
+		infra:  infra,
+	}
+}
+
+func (c *cloudProviderTests) run(t *testing.T) error {
+	if err := c.createPVCWithStorage(t); err != nil {
+		return err
+	}
+	if err := c.exposePVC(t); err != nil {
+		return err
+	}
+
+	return c.cleanUp(t)
```

Comment:
cleanUp should be a defer function, or even better a [testing.T.Cleanup](https://pkg.go.dev/testing#T.Cleanup), and it should be registered before any cloud-resource-producing functions are run.

Reply:
This should be fixed now.